shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueRunner.kt

```diff
@@ -0,0 +1,101 @@
+package io.customer.shared.tracking.queue
+
+import io.customer.shared.database.TrackingTask
+import io.customer.shared.database.TrackingTaskQueryHelper
+import io.customer.shared.sdk.meta.Workspace
+import io.customer.shared.tracking.api.*
+import io.customer.shared.tracking.api.model.*
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.tracking.model.Activity
+import io.customer.shared.tracking.model.Task
+import io.customer.shared.tracking.model.TaskResponse
+import io.customer.shared.util.JsonAdapter
+import io.customer.shared.util.Logger
+
+/**
+ * Class responsible for sending tasks to server by communicating with network layer and updating
+ * tasks status and responses in database.
+ */
+internal interface QueueRunner {
+    suspend fun runQueueForTasks(pendingTasks: List<TrackingTask>): Result<Boolean>
+}
+
+internal class QueueRunnerImpl(
+    private val logger: Logger,
+    private val jsonAdapter: JsonAdapter,
+    private val workspace: Workspace,
+    private val trackingTaskQueryHelper: TrackingTaskQueryHelper,
+    private val trackingHttpClient: TrackingHttpClient,
+) : QueueRunner {
+    override suspend fun runQueueForTasks(pendingTasks: List<TrackingTask>): Result<Boolean> {
+        return kotlin.runCatching {
+            trackingTaskQueryHelper.updateTasksStatus(
+                status = QueueTaskStatus.SENDING,
+                tasks = pendingTasks,
+            )
+            val response = processBatchTasks(pendingTasks = pendingTasks)
+            return@runCatching response != null && !response.isServerUnavailable
+        }.onFailure { ex ->
+            logger.error("Failed to run queue for ${pendingTasks.size} tasks with error: ${ex.message}")
+            trackingTaskQueryHelper.updateTasksStatus(
+                status = QueueTaskStatus.FAILED,
+                tasks = pendingTasks,
+            )
+        }
+    }
+
+    @Throws(Exception::class)
+    private suspend fun processBatchTasks(pendingTasks: List<TrackingTask>): BatchTrackingResponse? {
+        val activities = pendingTasks.map { task ->
+            jsonAdapter.fromJSON(Activity::class, task.activityJson)
```

Comment:
```suggestion
            jsonAdapter.fromJSON<Activity>(task.activityJson)
```

Instead of passing in the class as a parameter into the JsonAdapter, can we use generics?

Reply:
Unfortunately current implementation doesn't support it, mainly because `JsonAdapter` is an interface and we cannot use `reified` and `inline` there. But if we want to drop the interface here like current Android SDK, then we can achieve this to look cleaner.

Reply:
I just realized, we can still achieve this with the following extension, so I'm updating this
```
inline fun <reified T : Any> JsonAdapter.fromJSON(json: String): T = fromJSON(T::class, json)
```